flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/backpressure/BackPressureRequestCoordinatorTest.java

```diff
@@ -0,0 +1,311 @@
+import org.apache.flink.runtime.executiongraph.ExecutionVertex;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.messages.TaskBackPressureResponse;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.ArgumentMatchers;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for the {@link BackPressureRequestCoordinator}.
+ */
+public class BackPressureRequestCoordinatorTest extends TestLogger {
+
+	private static final long requestTimeout = 10000;
+	private static final double backPressureRatio = 0.5;
+
+	private static ScheduledExecutorService executorService;
+	private BackPressureRequestCoordinator coord;
+
+	@BeforeClass
+	public static void setUp() throws Exception {
+		executorService = new ScheduledThreadPoolExecutor(1);
+	}
+
+	@AfterClass
+	public static void tearDown() throws Exception {
+		if (executorService != null) {
+			executorService.shutdown();
+		}
+	}
+
+	@Before
+	public void init() throws Exception {
+		coord = new BackPressureRequestCoordinator(executorService, requestTimeout);
+	}
+
+	/** Tests simple request of task back pressure stats. */
+	@Test(timeout = 10000L)
+	public void testTriggerBackPressureRequest() throws Exception {
+		ExecutionVertex[] vertices = createExecutionVertices(ExecutionState.RUNNING, CompletionType.SUCCESSFULLY);
+
+		CompletableFuture<BackPressureStats> requestFuture = coord.triggerBackPressureRequest(vertices);
+		BackPressureStats backPressureStats = requestFuture.get();
+
+		// verify the request result
+		assertEquals(0, backPressureStats.getRequestId());
+		assertTrue(backPressureStats.getEndTime() >= backPressureStats.getStartTime());
+
+		Map<ExecutionAttemptID, Double> tracesByTask = backPressureStats.getBackPressureRatios();
+		for (ExecutionVertex executionVertex : vertices) {
+			ExecutionAttemptID executionId = executionVertex.getCurrentExecutionAttempt().getAttemptId();
+			assertEquals(backPressureRatio, tracesByTask.get(executionId), 0.0);
+		}
+
+		// verify no more pending request
+		assertEquals(0, coord.getNumberOfPendingRequests());
+
+		// verify no error on late collect
+		coord.collectTaskBackPressureStat(0, vertices[0].getCurrentExecutionAttempt().getAttemptId(), 0.0);
```

Comment:
We should re-arrange the relevant tests in this class. In the `createExecutionVertices` we define the different task state and the complete type, then we can only verify the results based on these conditions. The final verified results can be ratio value, pending requests, etc. 

But we do not need to expose the methods of `coord#collectTaskBackPressureStat` and `coord#cancelBackPressureRequest` to call from the tests, because these methods are already covered by handling the rpc response logic internally. We can verify the final results to cover them.

From coordinator aspect, only `coord#triggerBackPressureRequest` and coord#shutdown` are stable for interacting with outside world, and all the internal methods could be refactored to break/merge future. So it is not proper to call these internal methods. Also it seems strange to call them separately in the tests .